File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/repo/chat/ChatFileRepository.kt
package jp.co.toukei.log.trustar.repo.chat

import io.reactivex.rxjava3.core.Single
import jp.co.toukei.log.trustar.Current
import jp.co.toukei.log.trustar.rest.model.FileKey
import jp.co.toukei.log.trustar.rest.model.FileKey2
import okhttp3.MultipartBody
import okhttp3.RequestBody

class ChatFileRepository {

    fun uploadFile(filename: String, body: RequestBody): Single<FileKey> {
        return uploadFile2(filename, body).map { it.key }
    }

    fun uploadFile2(filename: String, body: RequestBody): Single<FileKey2> {
        val p = MultipartBody.Part.createFormData("buffer", filename, body)
        return Current.chatFileApi.uploadFile(p)
    }

    fun uploadImage(filename: String, body: RequestBody): Single<FileKey> {
        val p = MultipartBody.Part.createFormData("buffer", filename, body)
        return Current.chatFileApi.uploadImage(p)
    }

    fun uploadVideo(filename: String, body: RequestBody): Single<FileKey2> {
        return uploadFile2(filename, body)
    }

}

File: LogiOS/TrustarApp/Repo/Chat/ChatFileRepository.swift
//
//  ChatFileRepository.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/11/25.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import Combine
import MobileCoreServices

func mimeTypeForPath(path: String) -> String {
    let url = NSURL(fileURLWithPath: path)
    let pathExtension = url.pathExtension

    if let uti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, pathExtension! as NSString, nil)?.takeRetainedValue() {
        if let mimetype = UTTypeCopyPreferredTagWithClass(uti, kUTTagClassMIMEType)?.takeRetainedValue() {
            return mimetype as String
        }
    }
    return "application/octet-stream"
}

class ChatFileRepository {
    func uploadFile(filename: String, body: URL) -> AnyPublisher<FileKey, NetworkError> {
        return uploadFile2(filename: filename, body: body)
            .flatMap({ f2 in
                return Just(f2.key())
                    .setFailureType(to: NetworkError.self)
                    .eraseToAnyPublisher()
            })
            .eraseToAnyPublisher()
    }
    
    func uploadFile2(filename: String, body: URL) -> AnyPublisher<FileKey2, NetworkError> {
        return Current.Shared.chatFileApi.uploadFile(body: body, fileName: filename, apiKey: Current.Shared.loggedUser?.token ?? "")
    }
    
    func uploadImage(filename: String, body: URL) -> AnyPublisher<FileKey, NetworkError> {
        return Current.Shared.chatFileApi.uploadImage(body: body, fileName: filename, apiKey: Current.Shared.loggedUser?.token ?? "")
    }
    
    func uploadVideo(filename: String, body: URL) -> AnyPublisher<FileKey2, NetworkError> {
        return uploadFile2(filename: filename, body: body)
    }
}
